File: Cores/DotNet/Optimization/Nodes/TerminationViaMaxTime.cs
using System;
using System.Collections.Generic;

namespace OSOL.Extremum.Cores.DotNet.Optimization.Nodes
{

    using Area = Dictionary<string, Tuple<double, double>>;

    public class TerminationViaMaxTime<TBase, TFuncType, TV> : GeneralNode<TBase, TFuncType, TV>
        where TV : class, IOptimizable<TBase, TFuncType>
    {
        private readonly string ParameterName;
        private readonly double MaxTime;

        public TerminationViaMaxTime(int nodeId, double maxTime, string parameterName = "startTime")
        {
            this.NodeId = nodeId;
            this.MaxTime = maxTime;
            this.ParameterName = parameterName;
        }

        public override void Initialize(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area,
            State<TBase, TFuncType, TV> state)
        {
            DateTime? value = null;
            try
            {
                value = state.GetParameter<DateTime>(ParameterName);
            }
            catch (OptimizationExceptions.NoSuchParameterException e)
            {
                state.SetParameter<DateTime>(name: ParameterName, value: DateTime.Now);
            }
            finally
            {
                if (value.HasValue)
                {
                    throw new OptimizationExceptions.ParameterAlreadyExistsException(ParameterName);
                }
            }
        }

        public override void Process(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area,
            State<TBase, TFuncType, TV> state)
        {
        }

        public override int? GetCurrentCondition(Func<Dictionary<string, TFuncType>, TFuncType> f, Area area,
            State<TBase, TFuncType, TV> state) =>
            (DateTime.Now - state.GetParameter<DateTime>(ParameterName)).TotalSeconds > MaxTime ? 1 : 0;

    }
}

File: Cores/JVM/src/main/scala/OSOL/Extremum/Cores/JVM/Optimization/Nodes/TerminationViaMaxTime.scala
package OSOL.Extremum.Cores.JVM.Optimization.Nodes

import OSOL.Extremum.Cores.JVM.Optimization._
import OSOL.Extremum.Cores.JVM.Optimization.Exceptions._

class TerminationViaMaxTime[Base, FuncType, V <: Optimizable[Base, FuncType]]
(override val nodeId: java.lang.Integer, val maxTime: java.lang.Double, parameterName: String = "startTime")
  extends GeneralNode[Base, FuncType, V ](nodeId) {

  final override def initialize(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Unit = {
    try {
      val _ = state.getParameter[java.lang.Long](parameterName)
      throw new ParameterAlreadyExistsException(parameterName)
    }
    catch {
      case _: NoSuchParameterException => state.setParameter(parameterName, System.nanoTime())
      case _ => throw new ParameterAlreadyExistsException(parameterName)
    }
  }

  final override def process(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Unit = { }

  final override def getCurrentCondition(f: Map[String, FuncType] => FuncType, area: Area, state: State[Base, FuncType, V]): Option[java.lang.Integer] = {
    if (1e-9 * (System.nanoTime() - state.getParameter[java.lang.Long](parameterName)) > maxTime) Some(1)
    else Some(0)
  }

}
